mem/test_mprotect.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Phoenix-RTOS
+ *
+ * phoenix-rtos-test
+ *
+ * mprotect syscall tests
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Hubert Badocha
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <sys/mman.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include "unity_fixture.h"
+
+static long page_size;
+
+#define PAGES 4
+
+TEST_GROUP(test_mprotect);
+
+
+TEST_SETUP(test_mprotect)
+{
+	page_size = sysconf(_SC_PAGESIZE);
+}
+
+
+TEST_TEAR_DOWN(test_mprotect)
+{
+}
+
+
+TEST(test_mprotect, test_mprotect_singlecore)
+{
+	unsigned char *area = mmap(0, page_size * PAGES, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+	TEST_ASSERT_NOT_NULL(area);
+
+	for (int page = 0; page < PAGES; page++) {
+		area[page * page_size] = 0x42;
+	}
+
+	TEST_ASSERT_EQUAL(0, mprotect(area, page_size * PAGES, PROT_READ));
+
+	for (int page = 0; page < PAGES; page++) {
+		TEST_ASSERT_EQUAL(0x42, area[page * page_size]);
+	}
+
+	TEST_ASSERT_EQUAL(0, mprotect(area, page_size * PAGES, PROT_READ | PROT_WRITE));
+
+	for (int page = 0; page < PAGES; page++) {
+		area[(page * page_size) + 0x6] = 0x9;
+		TEST_ASSERT_EQUAL(0x9, area[(page * page_size) + 0x6]);
+	}
+
+	TEST_ASSERT_EQUAL(0, munmap(area, page_size * PAGES));
+}
+
+
+struct mprotect_data {
+	pthread_cond_t cond;
+	pthread_mutex_t mutex;
+	enum { PROTECT,
+		USE,
+		EXIT } state;
+	int prot;
+	int res;
+	unsigned char *area;
+	size_t size;
+};
+
+
+static void *mprotect_protect_thread(void *arg)
+{
+	struct mprotect_data *data = arg;
+
+	TEST_ASSERT_EQUAL(0, pthread_mutex_lock(&data->mutex));
+
+	while (data->state != EXIT) {
+		if (data->state == PROTECT) {
+			data->res = mprotect(data->area, data->size, data->prot);
+			data->state = USE;
+			pthread_cond_signal(&data->cond);
+		}
+		TEST_ASSERT_EQUAL(0, pthread_cond_wait(&data->cond, &data->mutex));
+	}
+
+	TEST_ASSERT_EQUAL(0, pthread_mutex_unlock(&data->mutex));
+
+	return NULL;
+}
+
+
+static int mprotect_in_thread(struct mprotect_data *data, int prot)
+{
+	data->prot = prot;
+	data->state = PROTECT;
+
+	TEST_ASSERT_EQUAL(0, pthread_cond_signal(&data->cond));
+
+	do {
+		TEST_ASSERT_EQUAL(0, pthread_cond_wait(&data->cond, &data->mutex));
+	} while (data->state != USE);
+
+	return data->res;
+}
+
+
+TEST(test_mprotect, test_mprotect_multicore)
```

Comment:
I am not sure about this test. We don't have any guarantees that threads will be assigned to different cores, also threads seem to be serialized by lock and cond, so aren't run in parallel anyway